File: tests/unit/format.test.ts
// XTRITIUM §8 — `1.234,56 ₺` and `GG/AA/YYYY`, in both languages, on every
// machine. These assertions are the convention written down: if one fails, the
// app is showing the maker someone else's number format.

import { describe, expect, it } from 'vitest'
import {
  currencySymbol,
  formatDate,
  formatFigure,
  formatMoneyText,
  parseDate,
  parseInput,
  toInput
} from '../../src/shared/format.js'
import {
  MONEY_DECIMALS,
  PUMP_DECIMALS,
  TANK_DECIMALS,
  toMoney,
  toPump,
  toTank
} from '../../src/shared/scaled.js'

describe('figures the family way', () => {
  it('groups with a dot and separates the decimal with a comma', () => {
    expect(formatFigure(toMoney(1234.56), MONEY_DECIMALS)).toBe('1.234,56')
    expect(formatFigure(toMoney(2160000), MONEY_DECIMALS)).toBe('2.160.000,00')
    expect(formatFigure(toMoney(11746), MONEY_DECIMALS)).toBe('11.746,00')
  })

  it('keeps small figures ungrouped and zero-padded', () => {
    expect(formatFigure(toMoney(0), MONEY_DECIMALS)).toBe('0,00')
    expect(formatFigure(toMoney(7.5), MONEY_DECIMALS)).toBe('7,50')
    expect(formatFigure(toTank(54), TANK_DECIMALS)).toBe('54,0')
  })

  it('puts the sign in front of the whole figure', () => {
    expect(formatFigure(toMoney(-1234.56), MONEY_DECIMALS)).toBe('-1.234,56')
  })

  it('groups a whole number with no decimals at all', () => {
    // Odometers are plain integers and still read as the maker expects.
    expect(formatFigure(19764, 0)).toBe('19.764')
    expect(formatFigure(370, 0)).toBe('370')
  })
})

describe('currency', () => {
  it('spells out the codes it knows', () => {
    expect(formatMoneyText(toMoney(11746), 'TRY')).toBe('11.746,00 ₺')
    expect(currencySymbol('USD')).toBe('$')
    expect(currencySymbol('try')).toBe('₺')
  })

  it('prints a code it does not know as itself', () => {
    // §8 asks the question once as free text; an unknown answer is still valid
    // money and must never be refused or silently turned into something else.
    expect(currencySymbol('AZN')).toBe('AZN')
    expect(formatMoneyText(toMoney(50), 'AZN')).toBe('50,00 AZN')
  })
})

describe('reading a figure back out of a form', () => {
  it('takes the family convention', () => {
    expect(parseInput('1.234,56', MONEY_DECIMALS)).toBe(toMoney(1234.56))
    expect(parseInput('2.160.000,00', MONEY_DECIMALS)).toBe(toMoney(2160000))
  })

  it('takes the other convention too, rather than refusing a paste', () => {
    expect(parseInput('1,234.56', MONEY_DECIMALS)).toBe(toMoney(1234.56))
  })

  it('reads a lone separator as the decimal point, either way round', () => {
    expect(parseInput('54,0', TANK_DECIMALS)).toBe(toTank(54))
    expect(parseInput('54.0', TANK_DECIMALS)).toBe(toTank(54))
  })

  it('reads a repeated separator as grouping', () => {
    expect(parseInput('1.234.567', 0)).toBe(1234567)
  })

  it('does not guess grouping from three digits — pump figures have three', () => {
    // 8.165 is a real price per litre. A heuristic that read it as 8165 would
    // multiply the maker's fuel bill by a thousand.
    expect(parseInput('8.165', PUMP_DECIMALS)).toBe(toPump(8.165))
    expect(parseInput('73,380', PUMP_DECIMALS)).toBe(toPump(73.38))
  })

  it('rounds rather than truncates, so no cent goes missing', () => {
    // 19.99 × 100 is 1998.9999999999998 in IEEE doubles.
    expect(parseInput('19,99', MONEY_DECIMALS)).toBe(1999)
    expect(parseInput('4,35', MONEY_DECIMALS)).toBe(435)
  })

  it('keeps the sign', () => {
    expect(parseInput('-19,99', MONEY_DECIMALS)).toBe(-1999)
  })

  it('returns null for anything that is not a figure', () => {
    // null, not zero: the form must tell "nothing entered" from "entered zero".
    for (const text of ['', '   ', 'abc', '12abc', '1,2,3', '-', ',', '.']) {
      expect(parseInput(text, MONEY_DECIMALS)).toBeNull()
    }
  })
})

describe('a figure typed, stored, and shown again', () => {
  it('is the same figure, for every field TRITIUM has', () => {
    const cases: Array<[number, number]> = [
      [toMoney(2160000), MONEY_DECIMALS],
      [toMoney(11746), MONEY_DECIMALS],
      [toMoney(0), MONEY_DECIMALS],
      [toMoney(-8664.5), MONEY_DECIMALS],
      [toPump(29.99), PUMP_DECIMALS],
      [toPump(73.38), PUMP_DECIMALS],
      [toTank(54), TANK_DECIMALS]
    ]

    for (const [scaled, decimals] of cases) {
      expect(parseInput(toInput(scaled, decimals), decimals)).toBe(scaled)
    }
  })

  it('shows an editable field without grouping, so nothing invites a misread', () => {
    expect(toInput(toMoney(1234.56), MONEY_DECIMALS)).toBe('1234,56')
    expect(toInput(toTank(54), TANK_DECIMALS)).toBe('54,0')
  })
})

describe('dates', () => {
  it('shows a stored date as GG/AA/YYYY', () => {
    expect(formatDate('2026-08-16')).toBe('16/08/2026')
    expect(formatDate('2025-04-25')).toBe('25/04/2025')
  })

  it('shows nothing for a date it cannot read, rather than something wrong', () => {
    expect(formatDate('')).toBe('')
    expect(formatDate('16/08/2026')).toBe('')
  })

  it('reads GG/AA/YYYY back into what the file stores', () => {
    expect(parseDate('16/08/2026')).toBe('2026-08-16')
    expect(parseDate(' 25/04/2025 ')).toBe('2025-04-25')
  })

  it('refuses a day the calendar does not have', () => {
    // Date.UTC would roll 31/02 forward into March and silently store the wrong
    // day; the round-trip check catches it instead.
    expect(parseDate('31/02/2026')).toBeNull()
    expect(parseDate('00/01/2026')).toBeNull()
    expect(parseDate('16/13/2026')).toBeNull()
    expect(parseDate('2026-08-16')).toBeNull()
    expect(parseDate('')).toBeNull()
  })

  it('accepts the leap day only in a leap year', () => {
    expect(parseDate('29/02/2024')).toBe('2024-02-29')
    expect(parseDate('29/02/2026')).toBeNull()
  })
})

File: android/app/src/test/kotlin/io/github/sudomegas/tritium/storage/FormatTest.kt
package io.github.sudomegas.tritium.storage

import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Test

class FormatTest {

    @Test
    fun `currencySymbol recognises the four presets, case-insensitively`() {
        assertEquals("₺", Format.currencySymbol("TRY"))
        assertEquals("$", Format.currencySymbol("usd"))
        assertEquals("€", Format.currencySymbol("Eur"))
        assertEquals("£", Format.currencySymbol("GBP"))
    }

    @Test
    fun `an unrecognised currency code prints as itself`() {
        assertEquals("XAG", Format.currencySymbol("XAG"))
    }

    @Test
    fun `formatFigure groups the whole part in threes, XTRITIUM §4_4 style`() {
        assertEquals("1.234,56", Format.formatFigure(123456L, 2))
        assertEquals("11.746,00", Format.formatFigure(1174600L, 2))
        assertEquals("54,0", Format.formatFigure(540L, 1))
    }

    @Test
    fun `formatFigure handles negatives and small values`() {
        assertEquals("-11.746,00", Format.formatFigure(-1174600L, 2))
        assertEquals("0,00", Format.formatFigure(0L, 2))
    }

    @Test
    fun `formatMoneyText appends the symbol, or stands alone with no currency`() {
        assertEquals("11.746,00 ₺", Format.formatMoneyText(1174600L, "TRY"))
        assertEquals("11.746,00", Format.formatMoneyText(1174600L, ""))
    }

    @Test
    fun `toInput never groups`() {
        assertEquals("54,0", Format.toInput(540L, Scaled.TANK_DECIMALS))
        assertEquals("11746,00", Format.toInput(1174600L, Scaled.MONEY_DECIMALS))
    }

    @Test
    fun `parseInput reads a comma or a dot as the decimal separator`() {
        assertEquals(1174600L, Format.parseInput("11746,00", 2))
        assertEquals(1174600L, Format.parseInput("11746.00", 2))
    }

    @Test
    fun `parseInput reads grouped figures with both separators`() {
        assertEquals(123456L, Format.parseInput("1.234,56", 2))
        assertEquals(123456L, Format.parseInput("1,234.56", 2))
    }

    @Test
    fun `parseInput does not mistake three pump decimals for grouping`() {
        // §4.3: pump figures carry exactly three decimals, so 8.165 must read
        // as eight point one six five, never as eight thousand one sixty-five.
        assertEquals(8165L, Format.parseInput("8.165", 3))
    }

    @Test
    fun `parseInput repeated grouping reads as thousands`() {
        assertEquals(123456700L, Format.parseInput("1.234.567", 2))
    }

    @Test
    fun `parseInput rejects nonsense`() {
        assertNull(Format.parseInput("", 2))
        assertNull(Format.parseInput("abc", 2))
        assertNull(Format.parseInput("1,2,3", 2))
    }

    @Test
    fun `parseInput is the inverse of toInput for a round trip`() {
        val original = 8165L
        val text = Format.toInput(original, Scaled.PUMP_DECIMALS)
        assertEquals(original, Format.parseInput(text, Scaled.PUMP_DECIMALS))
    }

    @Test
    fun `formatDate and parseDate round-trip the family convention`() {
        assertEquals("16.08.2026", Format.formatDate("2026-08-16"))
        assertEquals("2026-08-16", Format.parseDate("16.08.2026"))
    }

    @Test
    fun `formatDate on an unparseable value shows nothing`() {
        assertEquals("", Format.formatDate(""))
        assertEquals("", Format.formatDate("not-a-date"))
    }

    @Test
    fun `parseDate refuses a day the real calendar does not have`() {
        // 2026 is not a leap year — no 29 February.
        assertNull(Format.parseDate("29.02.2026"))
        assertNull(Format.parseDate("31.04.2026"))
    }

    @Test
    fun `parseDate refuses the wrong shape entirely`() {
        assertNull(Format.parseDate("2026-08-16"))
        assertNull(Format.parseDate(""))
    }

    @Test
    fun `todayIso reads the local calendar date, not a UTC one`() {
        val now = java.time.LocalDateTime.of(2026, 8, 16, 1, 0, 0)
        assertEquals("2026-08-16", Format.todayIso(now))
    }
}
